### PR TITLE
chore: upgrade to Go 1.23.1

### PR DESCRIPTION
**Pull request by rootulp**
## Description

Upgrade to Go 1.23.1 motivated by https://github.com/celestiaorg/celestia-app/pull/3847. I had to modify GoReleaser config because https://github.com/goreleaser/goreleaser/issues/5127.

Maintainers need to update their local golangic-lint binary to v1.61.0. See [golangci-lint install](https://golangci-lint.run/welcome/install/).

### CODE REVIEW

**Review by evan-forbes — APPROVED** [review]

**Review by evan-forbes — DISMISSED** [review]

**Review by ninabarbakadze — APPROVED** [review]
utAck

**Comment by rootulp** [comment]
There's a bug in Go 1.23 https://github.com/golang/go/issues/69312 that is impacting celestia-node so let's hold off on merging this until we identify if the same issue may hit us.

File: .github/workflows/lint.yml [hunk 1]
```diff
@@ -22,10 +22,9 @@ jobs:
             go.sum
       - uses: golangci/golangci-lint-action@v6.1.0
         with:
-          version: v1.59.1
+          version: v1.60.3
```
Comment (rach-id) [inline .github/workflows/lint.yml:25]:
```suggestion
          version: v1.61.0
```

File: Dockerfile [hunk 2]
```diff
@@ -4,7 +4,7 @@
 #
 # Separating the builder and runtime image allows the runtime image to be
 # considerably smaller because it doesn't need to have Golang installed.
-ARG BUILDER_IMAGE=docker.io/golang:1.22.6-alpine3.19
+ARG BUILDER_IMAGE=docker.io/golang:1.23.1-alpine3.19
```
Comment (rach-id) [inline Dockerfile:7]:
```suggestion
ARG BUILDER_IMAGE=docker.io/golang:1.23.1-alpine3.20
```

File: Makefile [hunk 3]
```diff
@@ -7,7 +7,9 @@ DOCKER_PROTO_BUILDER := docker run -v $(shell pwd):/workspace --workdir /workspa
 PROJECTNAME=$(shell basename "$(PWD)")
 HTTPS_GIT := https://github.com/celestiaorg/celestia-app.git
 PACKAGE_NAME          := github.com/celestiaorg/celestia-app/v3
-GOLANG_CROSS_VERSION  ?= v1.22.6
+# Before upgrading the GOLANG_CROSS_VERSION, please verify that a Docker image exists with the new tag.
+# See https://github.com/goreleaser/goreleaser-cross/pkgs/container/goreleaser-cross
+GOLANG_CROSS_VERSION  ?= v1.23
```
Comment (rach-id) [inline Makefile:12]:
```suggestion
GOLANG_CROSS_VERSION  ?= v1.23.1
```

File: README.md [hunk 4]
```diff
@@ -134,7 +134,7 @@ This repo contains multiple go modules. When using it, rename `go.work.example`
 
 ### Tools
 
-1. Install [golangci-lint](https://golangci-lint.run/welcome/install) 1.59.1
+1. Install [golangci-lint](https://golangci-lint.run/welcome/install) 1.60.3
```
Comment (rach-id) [inline README.md:137]:
```suggestion
1. Install [golangci-lint](https://golangci-lint.run/welcome/install) 1.61.0
```

File: docker/Dockerfile_txsim [hunk 5]
```diff
@@ -1,5 +1,5 @@
 # Stage 1: generate celestia-appd binary
-FROM --platform=$BUILDPLATFORM docker.io/golang:1.22.6-alpine3.19 as builder
+FROM --platform=$BUILDPLATFORM docker.io/golang:1.23.1-alpine3.19 as builder
```
Comment (rach-id) [inline docker/Dockerfile_txsim:2]:
```suggestion
FROM --platform=$BUILDPLATFORM docker.io/golang:1.23.1-alpine3.20 as builder
```

**Review by rach-id — COMMENTED** [review]
Added some more bumps in case we want to bump everything that is bumpable that's related to go version.

Btw, I already bumped core to 1.23.1. If the bug is infecting that too, we can revert ASAP

File: Makefile (hunk 3 above)
Comment (rootulp) [reply Makefile:12]: Thanks! When I made the PR, the Docker image for v1.23.1 didn't exist yet but now it does.

**Comment by rootulp** [comment]
Looks like https://github.com/golang/go/issues/69312 was fixed in Go 1.23.2. celestia-app doesn't invoke `timer.Stop` so I think we can proceed with this Go bump.

**Review by rach-id — APPROVED** [review]

**Review by evan-forbes — APPROVED** [review]

**Pull request merged**